File: core/src/main/java/y2k/joyreactor/common/PersistentMap.java
package y2k.joyreactor.common;

import y2k.joyreactor.platform.Platform;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.StringReader;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by y2k on 19/10/15.
 */
public class PersistentMap {

    private File cookieFile;
    private Map<String, String> map = new ConcurrentHashMap<>();

    public PersistentMap(String name) {
        cookieFile = new File(Platform.Instance.getCurrentDirectory(), name);

        try {
            initialize();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void initialize() throws IOException {
        if (!cookieFile.exists()) return;

        String data = IoUtils.readAll(cookieFile);
        BufferedReader reader = new BufferedReader(new StringReader(data));
        String line;
        while ((line = reader.readLine()) != null)
            map.put(line, reader.readLine());
    }

    public void flush() throws IOException {
        if (map.isEmpty()) return;

        StringBuilder cookie = new StringBuilder();
        for (String key : map.keySet())
            cookie.append(key).append("\n").append(map.get(key)).append("\n");

        IoUtils.writeAll(cookieFile, cookie.toString());
    }

    public synchronized void clear() {
        cookieFile.delete();
        map.clear();
    }

    public boolean isEmpty() {
        return map.isEmpty();
    }

    public PersistentMap put(String key, String value) {
        map.put(key, value);
        return this;
    }

    public String get(String key) {
        return map.get(key);
    }

    public Set<String> keySet() {
        return map.keySet();
    }
}

File: core/src/main/kotlin/y2k/joyreactor/common/PersistentMap.kt
package y2k.joyreactor.common

import y2k.joyreactor.platform.Platform
import java.io.File
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by y2k on 19/10/15.
 */
class PersistentMap(name: String) {

    private val cookieFile: File
    private val map = ConcurrentHashMap<String, String>()

    init {
        cookieFile = File(Platform.Instance.currentDirectory, name)
        if (cookieFile.exists())
            map.putAll(cookieFile.readLines().groupToPair())
    }

    fun flush() {
        if (map.isEmpty()) return

        val cookie = StringBuilder()
        for (key in map.keys)
            cookie.append(key).append("\n").append(map[key]).append("\n")

        cookieFile.writeText(cookie.toString())
    }

    @Synchronized fun clear() {
        cookieFile.delete()
        map.clear()
    }

    val isEmpty: Boolean
        get() = map.isEmpty()

    fun put(key: String, value: String): PersistentMap {
        map.put(key, value)
        return this
    }

    operator fun get(key: String): String {
        return map[key]!!
    }

    fun keySet(): Set<String> {
        return map.keys
    }
}